cdap/resource_gcs_artifact.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdap
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"regexp"
+
+	"cloud.google.com/go/storage"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+var bucketPathRE = regexp.MustCompile(`^gs://(.+)/(.+)$`)
+
+func resourceGCSArtifact() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGCSArtifactCreate,
+		Read:   resourceArtifactRead,
+		Delete: resourceArtifactDelete,
+		Exists: resourceArtifactExists,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The name of the artifact.",
+			},
+			"namespace": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of the namespace in which this resource belongs. If not provided, the default namespace is used.",
+				DefaultFunc: func() (interface{}, error) {
+					return defaultNamespace, nil
+				},
+			},
+			"bucket_path": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Path to GCS bucket object containing the spec, JAR and JSON.",
+			},
+			"version": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The version of the artifact.",
+			},
+		},
+	}
+}
+
+type artifactSpec struct {
+	Actions []*action `json:"actions"`
+}
+type action struct {
+	Type string `json:"type"`
+	Args []*struct {
+		Name  string `json:"name"`
+		Value string `json:"value"`
+	} `json:"arguments"`
+}
+
+func resourceGCSArtifactCreate(d *schema.ResourceData, m interface{}) error {
+	ctx := context.Background()
+	config := m.(*Config)
+
+	matches := bucketPathRE.FindStringSubmatch(d.Get("bucket_path").(string))
+	if len(matches) != 3 {
```

Comment:
why 3? I guess the spec, JAR and JSON. maybe add a doc above.

Reply:
Done.